src/transforms/pipelines/config.rs

```diff
@@ -42,65 +52,172 @@ impl Clone for PipelineConfig {
+            InnerTopologyTransform {
+                inputs: inputs.to_vec(),
+                inner: Box::new(self.clone()),
+            },
+        );
+        // TODO: actually call outputs fn
         result
             .outputs
-            .push((last_name, vec![Output::default(DataType::all())]));
-        // insert the filter if needed and return the next inputs
-        let mut next_inputs = if let Some(ref filter) = self.filter {
-            let mut conditions = IndexMap::new();
-            conditions.insert("success".to_string(), filter.to_owned());
-            let filter_name = name.join("filter");
-            result.inner.insert(
-                filter_name.clone(),
-                InnerTopologyTransform {
-                    inputs: inputs.to_vec(),
-                    inner: Box::new(RouteConfig::new(conditions)),
-                },
-            );
-            result.outputs.push((
-                filter_name.clone(),
-                vec![Output::default(DataType::all()).with_port(UNMATCHED_ROUTE)],
-            ));
-            vec![filter_name.port("success")]
-        } else {
-            inputs.to_vec()
-        };
-        // compound like
-        for (index, transform) in self.transforms.iter().enumerate() {
-            let step_name = name.join(index);
-            result.inner.insert(
-                step_name.clone(),
-                InnerTopologyTransform {
-                    inputs: next_inputs,
-                    inner: transform.to_owned(),
-                },
-            );
-            next_inputs = vec![step_name.id().to_string()];
-        }
-        //
+            .push((name.clone(), vec![Output::default(DataType::all())]));
         Ok(Some(result))
     }
 }
 
+#[derive(Clone)]
+struct Pipeline {
+    condition: Option<Condition>,
+    transforms: Vec<Box<dyn SyncTransform>>,
+    buf_in: TransformOutputsBuf,
+    buf_out: TransformOutputsBuf,
+}
+
+impl SyncTransform for Pipeline {
+    fn transform(&mut self, _event: Event, _output: &mut TransformOutputsBuf) {
+        // NOTE This is a bit of a wart in the SyncTransform API. We could
+        // consider splitting out another BatchTransform variant for
+        // transform_all and implementing the trait for SyncTransform instead.
+        unimplemented!()
+    }
+
+    fn transform_all(&mut self, events: EventArray, output: &mut TransformOutputsBuf) {
+        // A `Pipeline` is a compound of sub-transforms. That is, it's a
+        // transform that runs other transforms. To achieve this we gate all
+        // incoming Events by whether they match the pipeline condition or not
+        // and, if they do not, immediately output them. If the Event does match
+        // our condition we queue it up for further processing.
+        //
+        // Here our queue is the TransformOutputsBuf. In the next chunk of code
+        // that follows we do the aforementioned filtering and push into
+        // `self.buf_out`.
+        {
```

Comment:
This is formatted in a sub-block, which seems significant, but I'm at a loss as to what that gives us other than four more spaces per line.

Reply:
Ah, vestigial. I can remove that.

Reply:
Oh, er, related to my output questions [here](https://github.com/vectordotdev/vector/pull/12602#discussion_r871830637) probably. So, maybe.